### PR TITLE
compute loss if targets is present instead of checking for self.training

### PR DESCRIPTION
**Pull request by Delaunay**
Fix for #1574 

Why:
Current code assume that loss is only useful in training mode.
This is not true, users might want validation and test loss.

How:
1. We replace current logic of checking for training mode by checking if targets are provided.
If targets are provided then we assume users are interested in loss.

2. We always return a loss dictionary even if it is empty.
This makes the API consistent and users do not have to change code between eval and training mode

### CODE REVIEW

**Comment by Delaunay** [comment]
Merged the master branch to get the `torch.jit` changes as well.
Failing tests are the same as the master branch.

##  Todos

- [x] updated `references/detection/`
- [ ]  if we pass the targets, we don't compute the postprocessed predictions, but if we don't, then they are postprocessed.
   - [ ] ? adding a new argument to the model to postprocess the outputs.
